### ISSUE TITLE
Implement retry in readonly open

### ISSUE THREAD

**Issue by Ngoguey42** [mirage/irmin#2122]
Since #2119 we are able to detect if a control file is being read with a data race. These data races can be solved by sleeping and retrying. We are however not capable of distinguishing between data races and corrupted control files, this implies that the number of retry should be bounded to fallback on a `Corrupted_control_file` error.

Also, since the GC finalisation deletes old files, a data race is possible if a RO instance stalls between the read the control file and the opening of the files. The data race would materialise as a `No_such_file_or_directory` error.

Both these data races could be fixed at once by implementing a retry in the file manager.

**Comment by Athishpranav2003** [comment]
Correct me if i am wrong
https://github.com/mirage/irmin/blob/649657d2bca5a74d841c59a394cf2711a5139448/src/irmin-pack/unix/file_manager.ml#L733 This function is used to open files in read_only format. Since Dataraces is additional cause of error(which is temporary) we should have retry mechanism in this procedure in case the file exist but not in ready state.

**Comment by Athishpranav2003** [comment]
@kayceesrk I am not sure whom to ping for this :(. Pinging you for help

**Comment by kayceesrk** [comment]
@art-w @samoht are be the best to ping. Note that most folks are likely to be away in August (holiday month in France).

**Comment by Athishpranav2003** [comment]
I see
Fine
Anyways no harm is waiting until they get back :)

**Comment by Athishpranav2003** [comment]
so @art-w can we map the `Corrupted_control_file` normally and the rest of the errors we can do a sleep and rety?

**Comment by art-w** [comment]
Yes so any `No_such_file_or_directory` error that happens during step 2 (open suffix, prefix, dict, index) and step 3 (open lower) should retry the whole `open_ro`. I believe the other errors should only result in closing any successfully open files and propagating the error (not retry).

**Comment by Athishpranav2003** [comment]
@art-w so i was thinking about performing this. 
We can check the value of suffix, prefix, dict etc and if its pointing to no_such_file_or_directory we sleep(give up the computation) and awake to recursively retry the entire function. In order retry the entire function i was thinking of using exceptions in this manner 
```ocaml
let try_finalize f x finally y =
  let res = try f x with exn -> finally y; raise exn in
  finally y;
  res
```

Do u have any suggestions?

**Comment by art-w** [comment]
That makes sense, but you might want to check a tutorial on Lwt, which is the scheduler used by every `let*` in the code (you might see it as `let%lwt` in the tutorials, which is the same). Sadly the `try ... with ...` syntax doesn't work with Lwt, but it provides alternative like `Lwt.catch` or [`Lwt.finalize`](https://ocaml.org/p/lwt/5.7.0/doc/Lwt/index.html#val-finalize) to achieve the same.

Since the bug is almost impossible to replicate, it could help if you start by writing a unit test that creates a new irmin-pack repo, add a few things, and then remove one of the internal irmin files before trying to reopen the store. This should trigger the `No_such_file_or_directory` error and allow you to experiment with the catch-and-retry logic.

**Comment by Athishpranav2003** [comment]
Hmm ok

Will try the same,

Seems like small but tricky thing. i have some problem with type hinting so will resolve it to start with
